src/smt/seq_regex.cpp

```diff
@@ -124,6 +152,64 @@ namespace smt {
         th.propagate_lit(nullptr, 1, &lit, acc_lit);
     }
 
+    /**
+    * Gets an overapproximating regex s_approx for the input string expression s.
+    * such that for any valuation v(s) of s, v(s) in L(s_approx).
+    * If the overapproximation is trivial then dotstar is returned.
+    */
+    expr_ref seq_regex::get_overapprox_regex(expr* s) {
+        expr_ref s_to_re(re().mk_to_re(s), m);
+        expr_ref epsilon(re().mk_epsilon(m.get_sort(s)), m);
+        expr_ref dotstar(re().mk_full_seq(m.get_sort(s_to_re)), m);
+        if (m.is_value(s)) {
+            return s_to_re;
+        }
+        else {
+            expr_ref_vector es(m);
+            expr_ref s_approx = epsilon;
+            unsigned int n = 0;
+            if (str().is_concat(s)) {
+                str().get_concat(s, es);
+                n = es.size() - 1;
+                // make sure to simplify so that epsilons are eliminated in
+                // concatenations -- e.g. a sequence
+                // (x ++ "" ++ y ++ "") will be approximated by .*
+                for (unsigned i = n; i != (unsigned)(-1); i--) {
+                    expr_ref elem_i = get_overapprox_regex(es.get(i));
+                    if (i == n) {
+                        s_approx = elem_i;
+                    }
+                    else if (!re().is_epsilon(elem_i)) {
```

Comment:
I thought you wanted to also avoid concatenating dotstar

Reply:
For example, you could have

expr_ref result(m), last(m);
for (expr* e : es) {
    s_approx = get_overapprox_regex(e);
    if (!result) 
        result = s_approx;
     else if (last != dotstar || s_approx != dotstar)
        result = str.mk_concat(result, s_approx)
     last = s_approx;
}
}